composer/datasets/hparams.py

```diff
@@ -80,3 +85,85 @@ def initialize_object(self, batch_size: int, dataloader_hparams: DataloaderHpara
+        """Creates a :class:`DataLoader` or :class:`DataloaderSpec` for this dataset.
+
+        Parameters:
+            batch_size (int): The size of the batch the dataloader should yield. This batch size is
+                device-specific and already incorporates the world size.
+            dataloader_hparams (DataloaderHparams): The dataset-independent hparams for the dataloader
+
+        Returns:
+            Dataloader or DataSpec: The dataloader, or if the dataloader yields batches of custom types,
+            a :class:`DataSpec`.
+        """
+        pass
+
+
+@dataclass
+class JpgClsWebDatasetHparams(WebDatasetHparams, SyntheticHparamsMixin):
+    """Common functionality for (jpg, cls) WebDatasets.
+
+    Parameters:
+        dataset_s3_bucket (str): S3 bucket or root directory where dataset is stored.
+        dataset_name (str): Key used to determine where dataset is cached on local filesystem.
+        n_train_samples (int): Number of training samples.
+        n_val_samples (int): Number of validation samples.
+        height (int): Sample image height in pixels.
+        width (int): Sample image width in pixels.
+        n_classes (int): Number of output classes.
+        channel_means (list of float): Channel means for normalization.
+        channel_stds (list of float): Channel stds for normalization.
+    """
+
+    def initialize_object(self, batch_size: int, dataloader_hparams: DataloaderHparams) -> DataLoader:
+        if self.use_synthetic:
+            dataset = SyntheticBatchPairDataset(
+                total_dataset_size=self.n_train_samples if self.is_train else self.n_val_samples,
+                data_shape=[3, self.height, self.width],
+                num_classes=self.n_classes,
+                num_unique_samples_to_create=self.synthetic_num_unique_samples,
+                device=self.synthetic_device,
+                memory_format=self.synthetic_memory_format,
+            )
+        else:
+            if self.is_train:
+                split = 'train'
+                transform = transforms.Compose([
+                    transforms.RandomCrop((self.height, self.width), (self.height // 8, self.width // 8)),
+                    transforms.RandomHorizontalFlip(),
+                    transforms.ToTensor(),
+                    transforms.Normalize(self.channel_means, self.channel_stds),
+                ])
+            else:
+                split = 'val'
+                transform = transforms.Compose([
+                    transforms.ToTensor(),
+                    transforms.Normalize(self.channel_means, self.channel_stds),
+                ])
+            dataset, meta = load_webdataset(self.dataset_s3_bucket, self.dataset_name, split, self.webdataset_cache_dir,
+                                            self.webdataset_cache_verbose)
+            if self.shuffle:
+                dataset = dataset.shuffle(512)
```

Comment:
Can we have an hparam for the shuffle buffer size? Same for all the other webdatasets too. I think it may also be important to let the user know that the buffer size is **per worker**. 

I'm also curious how much this slows down the training init time? If each worker has to wait for 512 samples to arrive, that could be a good chunk of the dataset that needs to be downloaded before a single sample is returned. Just food for thought, and another reason why it could be good to have it as an hparam.

Reply:
Done